### PR TITLE
MULE-13939 Optimise AbstractEventContext callbacks for performance

### PR DESCRIPTION
**Pull request by dfeist**

### CODE REVIEW

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java [hunk 1]
```diff
@@ -37,98 +33,78 @@
...
 
-  private transient MonoProcessor<CoreEvent> beforeResponseProcessor;
-  private transient MonoProcessor<CoreEvent> responseProcessor;
-  private transient MonoProcessor<Void> completionProcessor;
-  private transient Disposable completionSubscriberDisposable;
   private transient final List<BaseEventContext> childContexts = new LinkedList<>();
-  private transient Mono<Void> completionCallback;
-  private transient FlowExceptionHandler exceptionHandler;
+  private transient final FlowExceptionHandler exceptionHandler;
+  private final MonoProcessor<CoreEvent> responseProcessor = MonoProcessor.create();
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java]: transient

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java [hunk 2]
```diff
@@ -37,98 +33,78 @@
...
   }
 
   public AbstractEventContext(FlowExceptionHandler exceptionHandler, Publisher<Void> completionCallback) {
-    this.completionCallback = from(completionCallback);
+    if (completionCallback != null) {
+      externalCompletion = false;
+      from(completionCallback).doOnSuccessOrError((aVoid, throwable) -> {
+        externalCompletion = true;
+        tryTerminate();
+      }).subscribe();
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java]: is it correct to subscribe this here? woudln't that execute it right now on this thread?

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java [hunk 3]
```diff
@@ -153,40 +129,101 @@ public final void success(CoreEvent event) {
...
-        responseProcessor.onError(throwable);
+        responseDone(null, throwable);
         return empty();
       }
     }
   }
 
+  private synchronized void responseDone(CoreEvent event, Throwable throwable) {
+    this.state = STATE_DONE;
+    onResponseConsumerList.stream().forEach(consumer -> consumer.accept(event, throwable));
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java]: what if one of these fail? Shouldn't the rest of this method be executed anyway? Or suppose you have 3 consumers, if one fails, shouldn't the rest be invoked also?

File: core/src/main/java/org/mule/runtime/core/privileged/processor/CompositeProcessorChainRouter.java [hunk 4]
```diff
@@ -59,10 +61,16 @@ protected ReactiveProcessor getExecutableFunction() {
...
   private BiFunction<CoreEvent, MessageProcessorChain, CoreEvent> processChain() {
     return (event, processorChain) -> {
+      Latch completionLatch = new Latch();
       BaseEventContext childContext = child((BaseEventContext) event.getContext(), ofNullable(getLocation()));
+      childContext.onTerminated((response, throwable) -> completionLatch.countDown());
       CoreEvent result = from(processWithChildContext(event, processorChain, childContext)).block();
       // Block until all child contexts are complete
-      from(childContext.getCompletionPublisher()).block();
+      try {
+        completionLatch.await();
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/privileged/processor/CompositeProcessorChainRouter.java]: no timeout?

File: core/src/main/java/org/mule/runtime/core/privileged/processor/CompositeProcessorChainRouter.java [hunk 5]
```diff
@@ -59,10 +61,16 @@ protected ReactiveProcessor getExecutableFunction() {
...
+      } catch (InterruptedException e) {
+        throw new MuleRuntimeException(e);
+      }
       return result;
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/privileged/processor/CompositeProcessorChainRouter.java]: add troubleshooting friendly message

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java [hunk 6]
```diff
@@ -153,40 +129,101 @@ public final void success(CoreEvent event) {
...
-  public Publisher<Void> getCompletionPublisher() {
-    return completionProcessor;
+  public synchronized void onTerminated(BiConsumer<CoreEvent, Throwable> consumer) {
+    onTerminatedConsumerList.add(consumer);
   }
 
-  protected FlowExceptionHandler getExceptionHandler() {
-    return exceptionHandler;
+  @Override
+  public synchronized void onComplete(BiConsumer<CoreEvent, Throwable> consumer) {
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java]: why do you add this but didn't change all the uses of `from(eventContext.getCompletionPublisher())`? For example, look at `org.mule.runtime.core.internal.streaming.CursorManager#registerEventContext`, shouldn't it be using this new method as well?... In either way, seems to me like now I have two ways of doing the same thing

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java [hunk 7]
```diff
@@ -153,40 +129,101 @@ public final void success(CoreEvent event) {
...
+  protected synchronized void tryComplete() {
+    if (this.state == STATE_DONE && childContexts.stream().filter(context -> !context.isComplete()).count() == 0) {
+      if (LOGGER.isDebugEnabled()) {
+        LOGGER.debug(this + " completed.");
+      }
+      this.state = STATE_COMPLETE;
+      if (responseProcessor.isError()) {
+        onCompletionConsumerList.forEach(runnable -> runnable.accept(null, responseProcessor.getError()));
+      } else {
+        onCompletionConsumerList.forEach(consumer -> consumer.accept(responseProcessor.peek(), null));
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java]: consumer vs runnable: use conssitent naming

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java [hunk 8]
```diff
@@ -153,40 +129,101 @@ public final void success(CoreEvent event) {
...
+  protected synchronized void tryTerminate() {
+    if (this.state == STATE_COMPLETE && externalCompletion) {
+      if (LOGGER.isDebugEnabled()) {
+        LOGGER.debug(this + " terminated.");
+      }
+      this.state = STATE_TERMINATED;
+      if (responseProcessor.isError()) {
+        onTerminatedConsumerList.forEach(runnable -> runnable.accept(null, responseProcessor.getError()));
+      } else {
+        onTerminatedConsumerList.forEach(consumer -> consumer.accept(responseProcessor.peek(), null));
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java]: consumer vs runnable: use conssitent naming

File: core/src/main/java/org/mule/runtime/core/internal/construct/AbstractPipeline.java [hunk 9]
```diff
@@ -267,18 +271,25 @@ protected ReactiveProcessor processFlowFunction() {
...
-          .doOnSuccessOrError((result, throwable) -> ((BaseEventContext) event.getContext()).getProcessingTime()
-              .ifPresent(time -> time.addFlowExecutionBranchTime(startTime)))
-          .subscribe(requestUnbounded());
+      BaseEventContext baseEventContext = ((BaseEventContext) event.getContext());
+      Reference<Pair<CoreEvent, Throwable>> result = new Reference<>();
+      baseEventContext.onResponse((response, throwable) -> {
+        result.set(new Pair<>(response, throwable));
+      });
+
+      baseEventContext.onComplete((a, b) -> {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/construct/AbstractPipeline.java]: receive the event and throwable here instead of using the previous block to get it.

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java (hunk 6 above)
Comment (dfeist) [reply core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java]: Did you not see these changes:
https://github.com/mulesoft/mule/pull/6147/files#diff-8ab0c830035b41ab12017e91072885bdR137
https://github.com/mulesoft/mule/pull/6147/files#diff-c7ba2cf2082917d74cf6465bc46d16bbL185

Comment (marianogonzalez) [reply core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java]: i clearly did not 😱

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java (hunk 2 above)
Comment (dfeist) [reply core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java]: It registers interest in this thread, which is fine.

File: core/src/main/java/org/mule/runtime/core/privileged/processor/CompositeProcessorChainRouter.java (hunk 4 above)
Comment (dfeist) [reply core/src/main/java/org/mule/runtime/core/privileged/processor/CompositeProcessorChainRouter.java]: no

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java [hunk 10]
```diff
@@ -37,98 +33,78 @@
  */
 abstract class AbstractEventContext implements BaseEventContext {
 
+  final static int STATE_READY = 0;
```
[the remark this answers is not in the document]
Comment (dfeist) [reply core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java]: `int` is used to do `>COMPLETE` etc.

[the remark this answers is not in the document]
Comment (marianogonzalez) [reply core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java]: you can do the same thing with the ordinal(). but i wouldn't use the enums either

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java [hunk 11]
```diff
@@ -137,13 +113,13 @@ public final void success(CoreEvent event) {
   @Override
   public final Publisher<Void> error(Throwable throwable) {
     synchronized (this) {
-      if (responseProcessor.isTerminated()) {
-        LOGGER.debug(this + " error response was already completed, ignoring.");
+      if (state > 0) {
+        LOGGER.debug(this + " error responseDone was already completed, ignoring.");
```
[the remark this answers is not in the document]
Comment (marianogonzalez) [reply core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java]: state > READY

File: core/src/main/java/org/mule/runtime/core/internal/construct/DefaultFlowBuilder.java [hunk 12]
```diff
@@ -253,7 +254,7 @@ public CoreEvent process(final CoreEvent event) throws MuleException {
               MessagingException me = new MessagingException(event, ree, this);
               ((BaseEventContext) request.getContext()).error(exceptionResolver.resolve(me, getMuleContext()));
             }
-            return Mono.from(((BaseEventContext) request.getContext()).getResponsePublisher())
+            return Mono.from(((BaseEventContext) event.getContext()).getResponsePublisher())
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/construct/DefaultFlowBuilder.java]: why this change? add a comment in code

File: core/src/main/java/org/mule/runtime/core/internal/context/notification/DefaultProcessorsTrace.java [hunk 13]
```diff
@@ -19,20 +21,20 @@
...
    * 
    * @param processorPath the path to mask as executed.
    */
-  public void addExecutedProcessors(String processorPath) {
+  public synchronized void addExecutedProcessors(String processorPath) {
     executedProcessors.add(processorPath);
   }
 
   @Override
   public List<String> getExecutedProcessors() {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/context/notification/DefaultProcessorsTrace.java]: this method has to be synchronized too

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java [hunk 14]
```diff
@@ -37,130 +37,147 @@
  */
 abstract class AbstractEventContext implements BaseEventContext {
 
+  private static final int STATE_READY = 0;
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java]: no enum then?

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java [hunk 15]
```diff
@@ -37,130 +37,147 @@
...
   }
 
   public AbstractEventContext(FlowExceptionHandler exceptionHandler) {
-    this(exceptionHandler, empty());
+    this(exceptionHandler, null);
   }
 
-  public AbstractEventContext(FlowExceptionHandler exceptionHandler, Publisher<Void> completionCallback) {
-    this.completionCallback = from(completionCallback);
+  public AbstractEventContext(FlowExceptionHandler exceptionHandler, CompletableFuture completableFuture) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java]: make the completableFuture optional, or use one that returns a null result instead of nulling the future.
Also, rename the param and add javadoc to have an idea what it is used for

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java [hunk 16]
```diff
@@ -37,130 +37,147 @@
...
    */
   @Override
   public final void success() {
-    synchronized (this) {
-      if (responseProcessor.isTerminated()) {
-        LOGGER.debug(this + " empty response was already completed, ignoring.");
-        return;
-      }
+    if (isResponseDone()) {
+      LOGGER.debug(this + " empty response was already completed, ignoring.");
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java]: check the log level before

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java [hunk 17]
```diff
@@ -37,130 +37,147 @@
...
    */
   @Override
   public final void success(CoreEvent event) {
-    synchronized (this) {
-      if (responseProcessor.isTerminated()) {
-        LOGGER.debug(this + " response was already completed, ignoring.");
-        return;
-      }
+    if (isResponseDone()) {
+      LOGGER.debug(this + " response was already completed, ignoring.");
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java]: check the log level before

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java [hunk 18]
```diff
@@ -37,130 +37,147 @@
...
    */
   @Override
   public final Publisher<Void> error(Throwable throwable) {
-    synchronized (this) {
-      if (responseProcessor.isTerminated()) {
-        LOGGER.debug(this + " error response was already completed, ignoring.");
-        return empty();
-      }
+    if (isResponseDone()) {
+      LOGGER.debug(this + " error response was already completed, ignoring.");
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java]: check the log level before

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java [hunk 19]
```diff
@@ -37,130 +37,147 @@
...
+    if (throwable instanceof MessagingException) {
       if (LOGGER.isDebugEnabled()) {
-        LOGGER.debug(this + " response completed with error.");
+        LOGGER.debug(this + " handling messaging exception.");
       }
+      return just((MessagingException) throwable)
+          .flatMapMany(exceptionHandler)
+          .doOnNext(handled -> success(handled))
+          .doOnError(rethrown -> responseDone(left(rethrown)))
+          .materialize()
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java]: do you really need this  `.materialize().then().toProcessor()`? why?

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java [hunk 20]
```diff
@@ -37,130 +37,147 @@
...
+    state = STATE_RESPONSE;
+    tryComplete();
+    onResponseConsumerList.stream().forEach(consumer -> signalConsumerSilently(consumer));
+  }
 
-      if (throwable instanceof MessagingException) {
-        if (LOGGER.isDebugEnabled()) {
-          LOGGER.debug(this + " handling messaging exception.");
+  protected synchronized void tryComplete() {
+    if (state == STATE_RESPONSE && !childContexts.stream().anyMatch(context -> !context.isComplete())) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java]: use another operation to avoid the double negation for readability (`noneMatch`?)

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java [hunk 21]
```diff
@@ -37,130 +37,147 @@
...
+      this.state = STATE_TERMINATED;
+      onTerminatedConsumerList.forEach(consumer -> signalConsumerSilently(consumer));
+    }
+  }
+
+  private void signalConsumerSilently(BiConsumer<CoreEvent, Throwable> consumer) {
+    try {
+      consumer.accept(result.getRight(), result.getLeft());
+    } catch (Throwable t) {
+      LOGGER.warn("The event consumer {}, of EventContext {} failed with exception: {} ", consumer, this, t);
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java]: why is this warn and not error?

File: core/src/main/java/org/mule/runtime/core/internal/event/DefaultEventContext.java [hunk 22]
```diff
@@ -139,7 +136,7 @@ public ProcessorsTrace getProcessorsTrace() {
    */
   public DefaultEventContext(FlowConstruct flow, ComponentLocation location,
                              String correlationId,
-                             Publisher<Void> externalCompletionPublisher) {
+                             CompletableFuture externalCompletionPublisher) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/event/DefaultEventContext.java]: rename this param

File: core/src/main/java/org/mule/runtime/core/internal/event/DefaultEventContext.java [hunk 23]
```diff
@@ -150,17 +147,18 @@ public DefaultEventContext(FlowConstruct flow, ComponentLocation location,
...
-  *        context, if available.
+   *        context, if available.
    * @param externalCompletionPublisher void publisher that completes when source completes enabling completion of
-  *        {@link BaseEventContext} to depend on completion of source.
+   *        {@link BaseEventContext} to depend on completion of source.
    * @param exceptionHandler the exception handler that will deal with an error context
    */
   public DefaultEventContext(String id, String serverId, ComponentLocation location, String correlationId,
-                             Publisher<Void> externalCompletionPublisher, FlowExceptionHandler exceptionHandler) {
+                             CompletableFuture<Void> externalCompletionPublisher, FlowExceptionHandler exceptionHandler) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/event/DefaultEventContext.java]: rename this param

File: core/src/main/java/org/mule/runtime/core/internal/event/DefaultEventContext.java [hunk 24]
```diff
@@ -185,7 +183,7 @@ public String toString() {
 
     private ChildEventContext(BaseEventContext parent, ComponentLocation componentLocation,
                               FlowExceptionHandler messagingExceptionHandler) {
-      super(messagingExceptionHandler, Mono.empty());
+      super(messagingExceptionHandler, null);
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/event/DefaultEventContext.java]: same as in https://github.com/mulesoft/mule/pull/6147/files#diff-c7ba2cf2082917d74cf6465bc46d16bbR66

File: core/src/main/java/org/mule/runtime/core/privileged/processor/CompositeProcessorChainRouter.java (hunk 4 above)
Comment (elrodro83) [reply core/src/main/java/org/mule/runtime/core/privileged/processor/CompositeProcessorChainRouter.java]: there should be a timeout, the shutdown timeout from the configuration in the muleContext

File: core-tests/src/test/java/org/mule/runtime/core/privileged/component/ExecutableComponentTestCase.java [hunk 25]
```diff
@@ -54,11 +54,10 @@ public void testExecuteWithInputEvent() throws Exception {
     assertThat(componentInEvent.get().getContext(), equalTo(response.getContext()));
 
     BaseEventContext eventContext = (BaseEventContext) componentInEvent.get().getContext();
-    assertThat(from(eventContext.getResponsePublisher()).toFuture().isDone(), is(true));
-    assertThat(from(eventContext.getCompletionPublisher()).toFuture().isDone(), is(false));
+    assertThat(eventContext.isTerminated(), is(false));
 
     executionResult.complete();
-    assertThat(from(eventContext.getCompletionPublisher()).toFuture().isDone(), is(true));
+    assertThat(eventContext.isTerminated(), is(true));
```
[the remark this answers is not in the document]
Comment (dfeist) [reply core-tests/src/test/java/org/mule/runtime/core/privileged/component/ExecutableComponentTestCase.java]: done. Added link in tests class javadoc.

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java (hunk 14 above)
Comment (dfeist) [reply core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java]: no. using an enum with ordinal is less clear and there is no real advantage.
